arch/arm/mach-rockchip/radxa_img.c

```diff
@@ -1,55 +1,145 @@
+#define VOLUME_LABEL32_OFFSET 0x47
+
+#define LABEL_LENGTH 11
+
+int get_partition_count(struct blk_desc *dev_desc) {
+	struct disk_partition info;
+	int part = 1;
+	int count = 0;
+
+	while (part_get_info(dev_desc, part, &info) == 0) {
+		count++;
+		part++;
+	}
+
+	return count;
+}
+
+int get_fat_volume_label(struct blk_desc *dev_desc, disk_partition_t *part_info, char *vol_label) {
+	void *buffer;
+	int is_fat32 = 0;
+
+	buffer = malloc_cache_aligned(dev_desc->blksz);
+	if (!buffer) {
+		printf("Error: Unable to allocate buffer\n");
+		return -ENOMEM;
+	}
+
+	if (blk_dread(dev_desc, part_info->start, 1, buffer) != 1) {
+		printf("Error: Unable to read boot sector\n");
+		free(buffer);
+		return -EIO;
+	}
+	unsigned char *buf = (unsigned char *)buffer;
+
+	if (buf[MBR_SIGNATURE_OFFSET] != 0x55 || buf[MBR_SIGNATURE_OFFSET + 1] != 0xAA) {
+		printf("Error: Invalid boot sector signature for partition.\n");
+		free(buffer);
+		return -EINVAL;
+	}
+
+	if (!memcmp(buffer + DOS_FS_TYPE_OFFSET, "FAT", 3)) {
+		is_fat32 = 0;
+	} else if (!memcmp(buffer + DOS_FS32_TYPE_OFFSET, "FAT32", 5)) {
+		is_fat32 = 1;
+	} else {
+		printf("Unknown filesystem type on partition.\n");
+		free(buffer);
+		return -EINVAL;
+	}
+
+	int offset = is_fat32 ? VOLUME_LABEL32_OFFSET : VOLUME_LABEL16_OFFSET;
+	memcpy(vol_label, buffer + offset, LABEL_LENGTH);
+	vol_label[LABEL_LENGTH] = '\0';
+
+	for (int i = LABEL_LENGTH - 1; i >= 0; i--) {
+		if (vol_label[i] == ' ') {
+			vol_label[i] = '\0';
+		} else {
+			break;
+		}
+	}
+
+	free(buffer);
+	return 0;
+}
 
 int radxa_read_bmp_file(void *buf, const char *name) {
 	struct blk_desc *desc = rockchip_get_bootdev();
-	disk_partition_t part;
+	disk_partition_t part_info;
 	loff_t actread, len;
+	int part_count, part_num;
+	char volume_label[LABEL_LENGTH + 1];
 
-	if (!desc)
-		return -ENODEV;
-
-	if (part_get_info_by_name(desc, "config", &part) < 0) {
-		printf("Failed to get info for config partition\n");
+	if (!desc) {
+		printf("Error: No boot device found.\n");
 		return -ENODEV;
 	}
 
-	if (fat_set_blk_dev(desc, &part) != 0) {
-		printf("Failed to set block device for FAT\n");
+	part_count = get_partition_count(desc);
+	if (part_count <= 0) {
+		printf("No partitions found.\n");
 		return -ENODEV;
 	}
 
-	if (fat_exists(name) != 1) {
-		printf("File %s does not exist\n", name);
-		return -ENOENT;
-	}
+	for (part_num = 1; part_num <= part_count; part_num++) {
+		if (part_get_info(desc, part_num, &part_info) != 0) {
+			printf("Failed to get information for partition %d\n", part_num);
+			continue;
+		}
 
-	if (fat_size(name, &len) < 0) {
-		printf("Failed to get file size for %s\n", name);
-		return -ENOENT;
-	}
+		if (get_fat_volume_label(desc, &part_info, volume_label) != 0) {
+			continue;
+		}
 
-	if (!buf) {
-		printf("Failed to allocate memory for file\n");
-		free(buf);
-		return -ENOMEM;
-	}
+		if ((strncmp(volume_label, "config", 6) == 0) || (strncmp(volume_label, "CONFIG", 6) == 0)) {
```

Comment:
```suggestion
		if (stricmp(volume_label, "config") == 0) {
```
你上面读取label的时候有末尾填0就应该可以不用指定字符串长度。这样可以保证比如`config1`不会匹配为成功

Reply:
u-boot 中没有 stricmp 的实现